Update Model Card for Mamba-2 (#37951)
huggingface/transformers: docs/source/en/model_doc/mamba2.md

-->

<div style="float: right;">
  <div class="flex flex-wrap space-x-1">
    <img alt="PyTorch" src="https://img.shields.io/badge/PyTorch-DE3412?style=flat&logo=pytorch&logoColor=white">
  </div>

# Mamba 2

> [Mamba2](https://huggingface.co/papers/2405.21060) is a selective structured state space model (SSMs) It brings many improvements to the original architecture such as parallelism, a larger head dimension and optimized state space model parameterization.

Reviewer: ```suggestion
[Mamba2](https://huggingface.co/papers/2405.21060) is the second iteration of selective structured state space model (SSMs) by Tri Dao and Albert Gu. It brings many improvements to the original architecture such as better parallelism support and more optimized support for higher dimensionalities.
```
Maybe something like this? The parametization is not really better tbh and the larger dimension corresponds to the state dim (so I left it vague here).

Reviewer: Maybe some description should be there that it's about codestral 👀

Author: I removed the description of codestral as it made the passage more complicated to understand. I have updated the description of the model and its architecture.

> You can find all the original Mamba checkpoints under the [State Space Models](https://huggingface.co/state-spaces) organization.

Author: I've given precedence to the codestral model and added it first. I think the original mamba2 checkpoints should be kept if anyone would like to experiment or open up PRs to improve compatibility for the model.

> [Mamba2](https://huggingface.co/papers/2405.21060) is the second iteration of selective structured state space model (SSMs) by Tri Dao and Albert Gu. It brings many improvements to the original architecture such as better parallelism support and more optimized support for higher dimensionalities. It is recommended to use the mistralai/Mamba-Codestral-7B-v0.1 model since it has the most stable performance.

Reviewer: ```suggestion
[Mamba2](https://huggingface.co/papers/2405.21060) is the second iteration of selective structured state space model (SSMs) by Tri Dao and Albert Gu. It brings many improvements to the original architecture such as better parallelism support and more optimized support for higher dimensionalities. 
```
My comments before were confusing sorry.

Author: Updated as discussed.

> You can Mamba2 based models such as [mistralai/Mamba-Codestral-7B-v0.1](https://huggingface.co/mistralai/Mamba-Codestral-7B-v0.1) model in the [mistral](https://huggingface.co/mistralai) organization and the original Mamba2 checkpoints under the [State Space Models](https://huggingface.co/state-spaces) organization.

Reviewer: Sorry to be annoying here but the original mamba2 checkpoints are not compatible with hf (needs some conversion). So let's remove the last part.

I agree that it would make sense to add them but since they're incompatible, it would only lead to confusion. 
```suggestion
Mamba2-based models such as [mistralai/Mamba-Codestral-7B-v0.1](https://huggingface.co/mistralai/Mamba-Codestral-7B-v0.1) can be found under the [mistral](https://huggingface.co/mistralai) organization.

Author: Updated as discussed.

> [!TIP]
> Click on the Mamba models in the right sidebar for more examples of how to apply Mamba to different language tasks.

The example below demonstrates how to generate text with [`Pipeline`], [`AutoModel`], and from the command line.

hfoptions id="usage">
<hfoption id="Pipeline">

```python
import torch
from transformers import pipeline

pipeline = pipeline(
    task="text-generation",

>     model="state-spaces/mamba2-130m",

Reviewer: This model won't work, since they need to be converted to be compatible with hf. (Same issue below then)

Let's stick to the codestral model for now.

Author: I've updated the examples to use the codestral model.

    torch_dtype=torch.float16,
    device=0
)
pipeline("Plants create energy through a process known as")
```

</hfoption>
<hfoption id="AutoModel">

```python
import torch  
from transformers import AutoModelForCausalLM, AutoTokenizer  

tokenizer = AutoTokenizer.from_pretrained("state-spaces/mamba2-130m")
model = AutoModelForCausalLM.from_pretrained("state-spaces/mamba2-130m", torch_dtype=torch.float16, device_map="auto")  
input_ids = tokenizer("Plants create energy through a process known as", return_tensors="pt").to("cuda")  

output = model.generate(**input_ids)  
print(tokenizer.decode(output[0], skip_special_tokens=True)
```

</hfoption>
<hfoption id="transformers CLI">

```bash
echo -e "Plants create energy through a process known as" | transformers run --task text-generation --model state-spaces/mamba2-130m --device 0
```

</hfoption>
</hfoptions>

> ## Notes

Reviewer: Did I see it right that the notes were one blob of text before? :D Nice job

Author: Thanks for noticing! Yes, I have updated the notes into readable points for understanding quickly.

> 

Author: I've added an example that uses TorchAo for quantization and also updated the toctree.

> [!TIP]
> Click on the Mamba models in the right sidebar for more examples of how to apply Mamba to different language tasks.

The example below demonstrates how to generate text with [`Pipeline`], [`AutoModel`], and from the command line.

hfoptions id="usage">
<hfoption id="Pipeline">

```python
import torch
from transformers import pipeline

pipeline = pipeline(
    task="text-generation",
    model="mistralai/Mamba-Codestral-7B-v0.1",
    torch_dtype=torch.bfloat16,
    device=0
)
pipeline("Plants create energy through a process known as")
```

</hfoption>
<hfoption id="AutoModel">

```python
import torch  
from transformers import AutoModelForCausalLM, AutoTokenizer  

tokenizer = AutoTokenizer.from_pretrained("mistralai/Mamba-Codestral-7B-v0.1")
model = AutoModelForCausalLM.from_pretrained("mistralai/Mamba-Codestral-7B-v0.1", torch_dtype=torch.bfloat16, device_map="auto")  
input_ids = tokenizer("Plants create energy through a process known as", return_tensors="pt").to("cuda")  

output = model.generate(**input_ids)  
print(tokenizer.decode(output[0], skip_special_tokens=True))
```

</hfoption>
<hfoption id="transformers CLI">

```bash

> echo -e "Plants create energy through a process known as" | transformers run --task text-generation --model mistralai/Mamba-Codestral-7B-v0.1 --device 0

Reviewer 2: Sorry, this should actually still be `transformers-cli` for the `run` command!

```suggestion
echo -e "Plants create energy through a process known as" | transformers-cli run --task text-generation --model mistralai/Mamba-Codestral-7B-v0.1 --device 0
```

```

</hfoption>
</hfoptions>

Quantization reduces the memory burden of large models by representing the weights in a lower precision. Refer to the [Quantization](../quantization/overview) overview for more available quantization backends.

The example below uses [torchao](../quantization/torchao) to only quantize the weights to 4-bit integers.

```py
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer, TorchAoConfig

> from torchao.quantization import Int4WeightOnlyConfig

Reviewer 2: ```suggestion
```

> quantization_config = Int4WeightOnlyConfig(group_size=128)
> quantization_config = TorchAoConfig(quant_type=quant_config)

Reviewer 2: ```suggestion
quantization_config = TorchAoConfig("int4_weight_only", group_size=128)
```
